File: DSA/Linkedlist/Code/DeletionInLinkedList.py
# Deletion in Linked List in Python

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedListDeletion:
    def __init__(self):
        self.head = None
    
    def add_node(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        temp = self.head
        while temp.next is not None:
            temp = temp.next
        temp.next = new_node
    
    def delete_at_position(self, pos):
        if self.head is None:
            return
        
        if pos == 1:
            self.head = self.head.next
            return
        
        temp = self.head
        for i in range(1, pos - 1):
            if temp is None:
                return
            temp = temp.next
        
        if temp is None or temp.next is None:
            return
        
        temp.next = temp.next.next
    
    def delete_by_value(self, value):
        if self.head is None:
            return
        
        if self.head.data == value:
            self.head = self.head.next
            return
        
        temp = self.head
        while temp.next is not None and temp.next.data != value:
            temp = temp.next
        
        if temp.next is not None:
            temp.next = temp.next.next
    
    def delete_all_occurrences(self, value):
        while self.head is not None and self.head.data == value:
            self.head = self.head.next
        
        if self.head is None:
            return
        
        temp = self.head
        while temp.next is not None:
            if temp.next.data == value:
                temp.next = temp.next.next
            else:
                temp = temp.next
    
    def display(self):
        temp = self.head
        elements = []
        while temp is not None:
            elements.append(str(temp.data))
            temp = temp.next
        print(" -> ".join(elements) + " -> NULL")

# Test
list1 = LinkedListDeletion()

print("=== Deletion in Linked List ===\n")

# Test 1: Delete at position
for i in range(1, 6):
    list1.add_node(i * 10)

print("Original list: ", end="")
list1.display()

list1.delete_at_position(2)
print("After deleting position 2: ", end="")
list1.display()

# Test 2: Delete by value
list1.delete_by_value(40)
print("After deleting value 40: ", end="")
list1.display()

print("\nComplexity Analysis:")
print("Delete at position: O(n)")
print("Delete by value: O(n)")
print("Delete all occurrences: O(n)")
print("Space: O(1)")

File: DSA/Linkedlist/Code/DeletionInLinkedList.java
// Deletion in Linked List in Java

class Node {
    int data;
    Node next;
    
    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

public class LinkedListDeletion {
    Node head;
    
    void addNode(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            return;
        }
        Node temp = head;
        while (temp.next != null) temp = temp.next;
        temp.next = newNode;
    }
    
    void deleteAtPosition(int pos) {
        if (head == null) return;
        
        if (pos == 1) {
            head = head.next;
            return;
        }
        
        Node temp = head;
        for (int i = 1; i < pos - 1 && temp != null; i++)
            temp = temp.next;
        
        if (temp == null || temp.next == null) return;
        
        temp.next = temp.next.next;
    }
    
    void deleteByValue(int value) {
        if (head == null) return;
        
        if (head.data == value) {
            head = head.next;
            return;
        }
        
        Node temp = head;
        while (temp.next != null && temp.next.data != value)
            temp = temp.next;
        
        if (temp.next != null)
            temp.next = temp.next.next;
    }
    
    void deleteAllOccurrences(int value) {
        while (head != null && head.data == value)
            head = head.next;
        
        if (head == null) return;
        
        Node temp = head;
        while (temp.next != null) {
            if (temp.next.data == value)
                temp.next = temp.next.next;
            else
                temp = temp.next;
        }
    }
    
    void display() {
        Node temp = head;
        while (temp != null) {
            System.out.print(temp.data + " -> ");
            temp = temp.next;
        }
        System.out.println("NULL");
    }
    
    public static void main(String[] args) {
        LinkedListDeletion list = new LinkedListDeletion();
        
        System.out.println("=== Deletion in Linked List ===\n");
        
        // Test 1: Delete at position
        for (int i = 1; i <= 5; i++)
            list.addNode(i * 10);
        
        System.out.print("Original list: ");
        list.display();
        
        list.deleteAtPosition(2);
        System.out.print("After deleting position 2: ");
        list.display();
        
        // Test 2: Delete by value
        list.deleteByValue(40);
        System.out.print("After deleting value 40: ");
        list.display();
        
        System.out.println("\nComplexity Analysis:");
        System.out.println("Delete at position: O(n)");
        System.out.println("Delete by value: O(n)");
        System.out.println("Delete all occurrences: O(n)");
        System.out.println("Space: O(1)");
    }
}
